3.0/m_relaymsg.cpp

```diff
@@ -0,0 +1,218 @@
+    {
+    }
+};
+
+// Handler for the @relaymsg message tag sent with forwarded PRIVMSGs
+class RelayMsgCapTag : public ClientProtocol::MessageTagProvider {
+private:
+    RelayMsgCap& cap;
+
+public:
+    RelayMsgCapTag(Module* mod, RelayMsgCap& Cap)
+        : ClientProtocol::MessageTagProvider(mod)
+        , cap(Cap)
+    {
+    }
+
+    bool ShouldSendTag(LocalUser* user, const ClientProtocol::MessageTagData& tagdata) CXX11_OVERRIDE
+    {
+        return cap.get(user);
+    }
+};
+
+// Handler for the RELAYMSG command (users and servers)
+class CommandRelayMsg : public Command {
+private:
+    RelayMsgCap& cap;
+    RelayMsgCapTag& captag;
+    std::bitset<UCHAR_MAX> invalid_chars_map;
+    std::string invalid_chars = "!+%@&#$:'\"?*,.";
+
+public:
+    std::string fake_host;
+    std::string fake_ident;
+
+    CommandRelayMsg(Module* parent, RelayMsgCap& Cap, RelayMsgCapTag& Captag)
+        : Command(parent, "RELAYMSG", 3, 3)
+        , cap(Cap)
+        , captag(Captag)
+    {
+        flags_needed = 'o';
+        syntax = "<channel> <nick> <text>";
+        allow_empty_last_param = false;
+
+        // Populate a map of disallowed nick characters. Based off m_chghost code
+        for (std::string::const_iterator n = invalid_chars.begin(); n != invalid_chars.end(); n++) {
+            invalid_chars_map.set(static_cast<unsigned char>(*n));
+        }
+    }
+
+    std::string GetFakeHostmask(std::string& nick) {
+        return InspIRCd::Format("%s!%s@%s", nick.c_str(), fake_ident.c_str(), fake_host.c_str());
+    }
+
+    CmdResult Handle(User* user, const CommandBase::Params& parameters)
+    {
+        std::string channame = parameters[0];
+        std::string nick = parameters[1];
+        std::string text = parameters[2];
+
+        // Check that the source has the relaymsg capability.
+        if (IS_LOCAL(user) && !cap.get(user)) {
+            return CMD_FAILURE;
+        }
+
+        Channel* channel = ServerInstance->FindChan(channame);
+        // Make sure the channel exists and the sender is in the channel
+        if (!channel)
+        {
+            user->WriteNumeric(Numerics::NoSuchChannel(channame));
+            return CMD_FAILURE;
+        }
+        if (!channel->HasUser(user))
+        {
+            user->WriteNumeric(Numerics::CannotSendTo(channel, "You must be in the channel to use this command."));
+            return CMD_FAILURE;
+        }
+
+        // Check that target nick is not already in use
+        if (ServerInstance->FindNick(nick))
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "RELAYMSG spoofed nick is already in use");
+            return CMD_FAILURE;
+        }
+
+        // Make sure the nick doesn't include any core IRC characters (e.g. *, !)
+        // This should still be more flexible than regular nick checking - in particular
+        // we want to allow "/" and "~" for relayers
+        for (std::string::const_iterator x = nick.begin(); x != nick.end(); x++)
```

Comment:
A better way to do this would be to get rid of the map and do:

```cpp
if (nick.find_first_of(""!+%@&#$:'\"?*,.";") != std::string::npos)
{
   // error
}
```